Add Mosaico Template Listing page based on Searchkit

monishdeb opened the pull request.

monishdeb committed:
Add MosacioTemplate APIv4 entity

monishdeb committed:
Add mosaico template category

monishdeb committed:
Add Searchkit files

monishdeb committed:
Add missing text filters

monishdeb committed:
Add required extension

colemanw commented:
I've reviewed the code and it looks good :+1:

mattwire commented:
@monishdeb With the optionvalues (categories) as managed entities is it possible to remove the default ones?

mattwire commented:
I think there will also be a problem that adding required extensions won't be picked up on upgrades so we'd also need a system check or similar to ensure that we enable those extensions on upgrade.

colemanw commented:
> I think there will also be a problem that adding required extensions won't be picked up on upgrades so we'd also need a system check or similar to ensure that we enable those extensions on upgrade.

IMO that problem is not limited to this extension, and the system-check should be added to core. However I don't think it should be a blocker to this PR because only the new functionality added in this PR will be affected by missing dependencies; existing functionality is unaffected.

@mattwire here is a core PR for that check: https://github.com/civicrm/civicrm-core/pull/22464

colemanw commented:
> I think there will also be a problem that adding required extensions won't be picked up on upgrades so we'd also need a system check or similar to ensure that we enable those extensions on upgrade.

Kind of. Managed entities can be modified but I'd have to double-check whether they can be deleted. IMO a good compromise here would be to only create one default option (Newsletters), delete the second option from this PR and let site admins configure it themselves.
@monishdeb

monishdeb committed:
Add missing category_id in install script, ext required check in system status

monishdeb commented:
@colemanw @mattwire I have made the following changes: 
1. Removed the 2nd option and kept only 'Newsletter' as the default option in the mgd file.
2. Changed new page label to 'Mosaico Templates (new)' from 'Mosaico Template List'
3. Added 5 actions Preview/Edit/Settings/Copy/Delete


The remaining items is - the action should trigger the ang events [here](https://github.com/veda-consulting-company/uk.co.vedaconsulting.mosaico/blob/2.x/ang/crmMosaico.crmstar/TemplateItem.html#L7L12)

colemanw committed:
Enable search display action buttons

Implements a custom template in the search display which embeds a component
which leverages the crmMosaicoTemplates service to perform the actions.

colemanw committed:
Add MosaicoTemplate.Cagegory field label

monishdeb committed:
Add configure icon for categories and update Mosaico Templates url

colemanw commented on ang/afsearchMosaicoTemplateList.aff.html:
```diff
@@ -2,7 +2,7 @@
     <af-field name="title" />
     <af-field name="base" />
-    <af-field name="category_id" defn="{input_attrs: {multiple: true}}" />
+    <af-field name="category_id" defn="{input_attrs: {multiple: true}}" /><a href='/civicrm/admin/options/mailing_template_category' title="{ts}Click here to configure the categories"><i class="crm-i fa-wrench" aria-hidden="true"></i></a>
```
@monishdeb that smarty tag isn't going to work in this tpl, need the angular syntax.
And it needs to have `target="crm-popup"`

colemanw committed:
Add dialog for creating template from new list view; delete old ListCtrl

Bumping version requirement because this depends on https://github.com/civicrm/civicrm-core/pull/22490

seamuslee001 commented on ang/crmMosaico/CreateFromBaseTemplate.html:
```diff
@@ -0,0 +1,28 @@
+    <input required class="form-control" id="new-tpl-title" ng-model="$ctrl.model.title">
+
+    <label for="new-tpl-category">{{:: ts('Category') }}</label>
+    <input class="form-control" id="new-tpl-category" ng-model="$ctrl.model.category" crm-entityref="{entity: 'OptionValue', api: {params: {option_group_id: 'mailing_template_category', is_active: 1}}, select: {allowClear:true, minimumInputLength: 0, placeholder: ts('Select')}}">
```
@colemanw @monishdeb I was having an issue with the category sticking and had to change this to be $ctrl.model.category_id here instead of category

colemanw replied:
Good catch @seamuslee001. Fixed.

colemanw committed:
Filter by category on New Mailing screen

lcdservices commented on CRM/Mosaico/Upgrader.php:
```diff
@@ -120,10 +120,27 @@ public function upgrade_4704() {
+  }
+  
+  /**
+   * Add menu for traditional mailing.
```
not a big deal, but the comment should be updated to reflect the change

colemanw replied:
Fixed.

colemanw committed:
Add css class to category selector

lcdservices commented:
This is working as expected. I think it's ready to be merged.

mattwire commented:
Merging based on feedback above. @colemanw @monishdeb if there is further work/fixes it can be done via follow up PRs before next release
